### PR TITLE
Sandwiches

### LINKED ISSUE #466: add sandwich support
maps api supports not only rendering cartodb layers but http and plain layers (see 2.3 here https://github.com/CartoDB/Windshaft/blob/master/doc/MapConfig-1.3.0.md)

layergroup in cartodb only supports cartodb layers, we should add api to support those layers.

The api should look like

```
 cartodb.createLayer(map, {
          user_name: 'examples',
          type: 'cartodb',
          sublayers: [
          {
                type: "http",
                urlTemplate: "http://{s}.basemaps.cartocdn.com/light_nolabels/{z}/{x}/{y}.png"
                //subdomains: [ "a", "b", "c" ]
          },
          {
             sql: 'select * from country_boundaries',
             cartocss: '#layer { polygon-fill: #F00; polygon-opacity: 0.3; line-color: #F00; }'
             }, {
               type: 'http',
               //urlTemplate: 'http://a.basemaps.cartocdn.com/light_only_labels/{z}/{x}/{y}.png',
               urlTemplate: 'http://{s}.tile.stamen.com/toner-labels/{z}/{x}/{y}.png',
               subdomains: [ "a", "b", "c" ]
             }]
        })
        .addTo(map)
        .done(function(layer) {
          setTimeout(function() {
            layer.getSubLayer(0).setUrlTemplate(
               'http://{s}.tile.stamen.com/toner/{z}/{x}/{y}.png'
            )
            layer.getSubLayer(2).hide();
          }, 3000)

```

I created a draft here to test the implementation but it does need to be used: https://github.com/CartoDB/cartodb.js/pull/445

### PR DESCRIPTION
**Pull request by alonsogarciapablo**
image

Fixes #466.

**Some highlights of the changes in cartodb.js:**
- cartodb.js now uses metadata from Windshaft to determine what layers should be rendered in the pngs. By default, non-torque layers are rendered.
- cartodb.createLayer now accepts a `filter` option to determine the types of layers that will be rendered.
- cartodb.js exposes different types of sublayers through `layer.getSubLayer(index)` (mapnik and http sublayers) to work with the map (layerDefinition in the code).

**Other work in this PR includes:**
- Updated the docs with:
  - Examples of how to create maps with multiple types of layers.
  - new API to work with sublayers.
  - `filter` parameter.
- Created [this example](https://github.com/CartoDB/cartodb.js/blob/sandwiches/examples/leaflet_sandwich.html) to test that everything works.
- Tried to leave `layer_definition.js` a bit more cleaner than I found it by renaming/extracting some methods (eg: getLayerToken -> createMap). `MapBase` is still doing too much IMHO (provides methods to communicate with the layer and also methods to work with a map/layerDefinition).
- Revisited/updated/rewrote some tests in `layer_definition.spec.js` to make them easier to follow.

Some PRs that happened before this beast:

- [x] [Sandwiches using map properties from tiler](https://github.com/CartoDB/cartodb.js/pull/507)
- [x] [Fix interactivity of named_maps with multiple types of layers and refactor](https://github.com/CartoDB/cartodb.js/pull/528)

@javisantana has been very helpful with the code reviews, but we need some more feedback from @CartoDB/frontend . What do you guys think? Thanks (and apologies for the massive PR)!

### CODE REVIEW

**Comment by viddo** [comment]
Just confirming that I had a look, but I'm afraid I don't know these parts enough to really give and good feedback. Code-wise looks OK

**Comment by xavijam** [comment]
Taking a look

**Comment by xavijam** [comment]
Superb work dude

File: examples/leaflet_sandwich.html [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+        // add cartodb layer with multiple sublayers of different types
+        cartodb.createLayer(map, {
+          user_name: 'examples',
+          type: 'cartodb',
+          sublayers: [
+            {
+              type: "http",
+              urlTemplate: "http://{s}.basemaps.cartocdn.com/dark_nolabels/{z}/{x}/{y}.png",
+              subdomains: [ "a", "b", "c" ]
```
Comment (xavijam) [inline examples/leaflet_sandwich.html]: Is attribution a valid parameter for http sublayers?

Comment (alonsogarciapablo) [inline examples/leaflet_sandwich.html]: These are valid attributes for http sublayers:
https://github.com/CartoDB/Windshaft/blob/master/doc/MapConfig-1.3.0.md#23-http-layers-options

Comment (xavijam) [inline examples/leaflet_sandwich.html]: Uhm, ok, couldn't we manage in the frontend side to add that attribution to the map?

**Comment by alonsogarciapablo** [comment]
Thanks guys!

@ohasselblad this PR includes [some changes in the docs](https://github.com/CartoDB/cartodb.js/pull/508/files#diff-8be2f642524623d5012c7b2cf4fa4eb9L300). Could you please review them? Thanks!

File: examples/leaflet_sandwich.html (hunk 1 above)
Comment (alonsogarciapablo) [inline examples/leaflet_sandwich.html]: That's a valid concern. Some http layers might require to add attributions to the map, so perhaps that's something we need to add to `cartodb.createLayer`... @javisantana what are your thoughts?

**Comment by ohasselblad** [comment]
docs look great to me @alonsogarciapablo, I didn't find anything to change :+1: I like the re-org!

**Comment by alonsogarciapablo** [comment]
Thanks @ohasselblad!

**Pull request merged**